### DISCUSSION TITLE
MPD on RasPi in Read-Only mode

### DISCUSSION THREAD

**Discussion by oe8bck** [MusicPlayerDaemon/MPD#1363] [Q&A] [answered]
Hi,

For my new bar in the ground-floor I would like to setup a Raspberry Pi with MPD as a kind of music box. It basically works, but now I want to set the SD-card to read-only in order not to have to change it regularilly. I followed [these instructions](https://medium.com/swlh/make-your-raspberry-pi-file-system-read-only-raspbian-buster-c558694de79) but the MPD daemon does not start.

```
pi@barpi(ro):~$ sudo mpd --verbose --no-daemon
config_file: loading file /etc/mpd.conf
Nov 29 14:57 : exception: Failed to bind to '[::]:6600'; Failed to bind socket: Address already in use
```

I also added a symlink from `/var/lib/mpd/` to `/tmp` and changed the `log_file` option in `mpd.conf` to `/var/log/mpd.log` which is also mounted with `tmnpfs`.

What could prevent MPD starting on a read-only file system, especially regarding the sockets?

Many thanks for any answer!
KR,
Christof

**Comment by MaxKellermann** [comment]
What's the point of making `/var/lib/mpd/` point to a `tmpfs`? Why not just ask MPD not to write anything to `/var/lib/mpd/` instead?

> but the MPD daemon does not start.

Because another process already occupies port 6600, as you can see in the error message. This has nothing to do with read-only.

> What could prevent MPD starting on a read-only file system, especially regarding the sockets?

This question is too generic, and giving a full reply would not be very helpful, but would consume lots of my time. Not a good investment. Let's talk about problems you encounter. So far, you did not encounter any (read-only related) problem.

**Comment by oe8bck** [comment]
Thanks for the very fast reply!

Sorry for the stupid question, how to say to MPD not to write anything to `/var/lib/mpd`? Is there an option to do so or do you mean to re-confifure everything in `mpd.conf` to be on a `tmpfs` directory?

BTW, I did not find any process occupying :6600 (sorry to mention this in my original question)
```
pi@barpi(ro):~$ sudo ss -tulpn
Netid     State      Recv-Q     Send-Q         Local Address:Port            Peer Address:Port     Process
udp       UNCONN     0          0                    0.0.0.0:41421                0.0.0.0:*         users:(("avahi-daemon",pid=246,fd=14))
udp       UNCONN     0          0                    0.0.0.0:5353                 0.0.0.0:*         users:(("avahi-daemon",pid=246,fd=12))
udp       UNCONN     0          0                    0.0.0.0:68                   0.0.0.0:*         users:(("dhcpcd",pid=280,fd=13))
udp       UNCONN     0          0                    0.0.0.0:631                  0.0.0.0:*         users:(("cups-browsed",pid=347,fd=7))
udp       UNCONN     0          0                          *:5353                       *:*         users:(("avahi-daemon",pid=246,fd=13))
udp       UNCONN     0          0                          *:47648                      *:*         users:(("avahi-daemon",pid=246,fd=15))
tcp       LISTEN     0          128                  0.0.0.0:22                   0.0.0.0:*         users:(("sshd",pid=348,fd=3))
tcp       LISTEN     0          128                127.0.0.1:631                  0.0.0.0:*         users:(("cupsd",pid=318,fd=8))
tcp       LISTEN     0          128                     [::]:22                      [::]:*         users:(("sshd",pid=348,fd=4))
tcp       LISTEN     0          128                    [::1]:631                     [::]:*         users:(("cupsd",pid=318,fd=7))
pi@barpi(ro):~$ fuser 6600/tcp
pi@barpi(ro):~$
```

Of course the question is very generic, I just wanted to ask if somebody has an idea what's going wrong. To be honest, I thought there are already folks out who are using a Raspberry Pi in read-only mode.

KR,
Ch.

**Reply by MaxKellermann** [reply]
https://www.musicpd.org/help/

**Reply by MaxKellermann** [reply]
> I just wanted to ask if somebody has an idea what's going wrong

But your question had nothing to do with what went wrong.

> Sorry for the stupid question, how to say to MPD not to write anything to /var/lib/mpd?

This question confuses me, because MPD doesn't know about `/var/lib/mpd`. It will only write anything there if you tell it to. So .... just don't tell it to.

**Reply by oe8bck** [reply]
OK, another try, this time with alll the points from 'Getting Help'

### mpd --version
see attached file [mpd--version.txt](https://github.com/MusicPlayerDaemon/MPD/files/7619268/mpd--version.txt)
### configuration file (mpd.conf)
see attached file [mpd.conf.txt](https://github.com/MusicPlayerDaemon/MPD/files/7619269/mpd.conf.txt)
### log file (--verbose)
```
pi@barpi(ro):~$ sudo mpd --verbose
config_file: loading file /etc/mpd.conf
Nov 29 15:31 : exception: Failed to bind to '[::]:6600'; Failed to bind socket: Address already in use
pi@barpi(ro):~$
```

### what do I expect MPD to do, and what is actually happening
I expect that it starts up and it does not. There is nobody occupying port :6600
```
pi@barpi(ro):~$ sudo ss -tulpn|grep :6600
pi@barpi(ro):~$
```

So why does the bind fail?

KR,
Christof

**Reply by oe8bck** [reply]
BTW, when I change the log level in mpd.conf to "verbose", the file `/var/log/mpd.log` is still empty!

**Reply by MaxKellermann** [reply]
From your `mpd.conf`:
```
bind_to_address                 "localhost"
bind_to_address                 "any"
```
First you bind to `localhost:6600` and then to `any:6600`, which conflicts with the former.

**Reply by MaxKellermann** [reply]
> This question confuses me, because MPD doesn't know about `/var/lib/mpd`. It will only write anything there if you tell it to. So .... just don't tell it to.

About this one: you told MPD to write a database file, playlists, a state file and a sticker database to `/var/lib/mpd`. Why did you do that?

**Comment by oe8bck** [comment, marked as answer]
Well, I got it running!

What did I do? I started from scratch, installed the latest Bullseye release, added (the latest) mpd and mpc, configured my DAC board, added some control logic in order to be able to control the music box by keypress. And when it working as I expected, I made the SD card read-only with `raspi-config` "Performance Options"->"Overlay File System".

@MaxKellermann you were right, my question was not related to MPD, it was a problem with the configuration of the Raspberry Pi Bullseye system. I have to admit, I did not think of a read-only system for years.

Thanks a lot, KR
Christof
